hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -730,7 +730,44 @@ def examples(self):
     def all_block_bounds(self):
         return self.shrink_target.all_block_bounds()
 
-    def pass_to_descendant(self):
+    @derived_value
+    def examples_by_label(self):
+        """An index of all examples grouped by their label, with
+        the examples stored in their normal index order."""
+
+        examples_by_label = defaultdict(list)
+        for ex in self.examples:
+            examples_by_label[ex.label].append(ex)
+        return dict(examples_by_label)
+
+    def calculate_descents(self):
+        """Returns a list of all pairs (i, j) such that
+        self.examples[i] is an ancestor of self.examples[j] and
+        they have the same label.
+        """
+        result = []
+
+        for ls in self.examples_by_label.values():
+            if len(ls) <= 1:
+                continue
+
+            for i, ex in enumerate(ls[:-1]):
+                hi = len(ls)
+                lo = i + 1
+                if ls[lo].start >= ex.end:
+                    continue
+                while lo + 1 < hi:
+                    mid = (lo + hi) // 2
+                    if ls[mid].start >= ex.end:
+                        hi = mid
+                    else:
+                        lo = mid
+                descents = ls[i + 1 : hi]
```

Comment:
If a “descent” is a pair of ancestor/descendant, then this list should be named `descendants` instead, since the descent pairs haven't been formed until the following line.